adm_program/installation/update_functions.php

```diff
@@ -0,0 +1,262 @@
+    }
+}
+
+/**
+ * @param bool $enable
+ */
+function toggleForeignKeyChecks($enable)
+{
+    global $gDbType, $gDb;
+
+    if ($gDbType === Database::PDO_ENGINE_MYSQL)
+    {
+        // disable foreign key checks for mysql, so tables can easily deleted
+        $sql = 'SET foreign_key_checks = ' . (int) $enable;
+        $gDb->queryPrepared($sql);
+    }
+}
+
+/**
+ * @param int $versionMain
+ * @param int $versionMinor
+ * @param int $versionPatch
+ */
+function doVersion2Update(&$versionMain, &$versionMinor, &$versionPatch)
+{
+    global $gLogger, $gDb, $gL10n, $g_tbl_praefix, $gSettingsManager;
+
+    // nun in einer Schleife alle Update-Scripte fuer Version 2 einspielen
+    while ($versionMain === 2)
+    {
+        $flagNextVersion = false;
+
+        // version 2 Admidio had sql and php files where the update statements where stored
+        // these files must be executed
+
+        // in der Schleife wird geschaut ob es Scripte fuer eine patch-version (3. Versionsstelle) gibt
+        // patch-version 0 sollte immer vorhanden sein, die anderen in den meisten Faellen nicht
+        for ($tmpPatchVersion = $versionPatch; $tmpPatchVersion <= 14; ++$tmpPatchVersion)
+        {
+            // output of the version number for better debugging
+            $gLogger->notice('UPDATE: Start executing update steps to version '.$versionMain.'.'.$versionMinor.'.'.$tmpPatchVersion);
+
+            $dbScriptsPath = ADMIDIO_PATH . '/adm_program/installation/db_scripts/';
+            $version = $versionMain . '_' . $versionMinor . '_' . $tmpPatchVersion;
+            $sqlFileName = 'upd_' . $version . '_db.sql';
+            $phpFileName = 'upd_' . $version . '_conv.php';
+
+            if (is_file($dbScriptsPath . $sqlFileName))
+            {
+                $sqlQueryResult = querySqlFile($gDb, $sqlFileName);
+
+                if ($sqlQueryResult === true)
+                {
+                    $flagNextVersion = true;
+                }
+                else
+                {
+                    showNotice($sqlQueryResult, 'update.php', $gL10n->get('SYS_BACK'), 'layout/back.png', true);
+                    // => EXIT
+                }
+            }
+
+            $phpUpdateFile = $dbScriptsPath . $phpFileName;
+            // check if an php update file exists and then execute the script
+            if (is_file($phpUpdateFile))
+            {
+                require($phpUpdateFile);
+                $flagNextVersion = true;
+            }
+
+            $gLogger->notice('UPDATE: Finish executing update steps to version '.$versionMain.'.'.$versionMinor.'.'.$tmpPatchVersion);
+        }
+
+        // keine Datei mit der patch-version gefunden, dann die Main- oder minor-version hochsetzen,
+        // solange bis die aktuelle Versionsnummer erreicht wurde
+        if (!$flagNextVersion && version_compare($versionMain . '.' . $versionMinor . '.' . $tmpPatchVersion, ADMIDIO_VERSION, '<'))
```

Comment:
Why do you only update $versionMinor if $flagNextVersion == false?
I think this is the reason for the loop within version 2.

Reply:
this is not my code. look here: https://github.com/Admidio/admidio/pull/687/files/7385e37afaa25d593ae971c012a29c52d8289d20#diff-c27e50483e21678ade212f20f1716125L422

Reply:
hmm strange, within the new functions this won't work.